sentry-android-replay/src/main/java/io/sentry/android/replay/ScreenshotRecorder.kt

```diff
@@ -0,0 +1,148 @@
+package io.sentry.android.replay
+
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.Color
+import android.graphics.Paint
+import android.graphics.RectF
+import android.os.Handler
+import android.os.HandlerThread
+import android.os.Looper
+import android.os.SystemClock
+import android.util.Log
+import android.view.PixelCopy
+import android.view.View
+import android.view.ViewGroup
+import android.view.ViewTreeObserver
+import io.sentry.android.replay.video.SimpleVideoEncoder
+import io.sentry.android.replay.viewhierarchy.ViewHierarchyNode
+import java.lang.ref.WeakReference
+import java.util.WeakHashMap
+import kotlin.system.measureTimeMillis
+
+internal class ScreenshotRecorder(
+    val rootView: WeakReference<View>,
+    val encoder: SimpleVideoEncoder
+) : ViewTreeObserver.OnDrawListener {
+
+    private val thread = HandlerThread("SentryReplay").also { it.start() }
+    private val handler = Handler(thread.looper)
+    private val bitmapToVH = WeakHashMap<Bitmap, ViewHierarchyNode>()
+
+    companion object {
+        const val TAG = "ScreenshotRecorder"
+    }
+
+    private var lastCapturedAtMs: Long? = null
+    override fun onDraw() {
+        // TODO: replace with Debouncer from sentry-core
+        val now = SystemClock.uptimeMillis()
+        if (lastCapturedAtMs != null && (now - lastCapturedAtMs!!) < 1000L) {
+            return
+        }
+        lastCapturedAtMs = now
+
+        val root = rootView.get()
+        if (root == null || root.width <= 0 || root.height <= 0) {
+            return
+        }
+
+        val window = root.phoneWindow ?: return
+        val bitmap = Bitmap.createBitmap(
+            root.width,
+            root.height,
+            Bitmap.Config.ARGB_8888
+        )
+        Log.e("BITMAP CREATED", bitmap.toString())
+
+        val time = measureTimeMillis {
+            val rootNode = ViewHierarchyNode.fromView(root)
+            root.traverse(rootNode)
+            bitmapToVH[bitmap] = rootNode
+        }
+        Log.e("TIME", time.toString())
+
+//    val latch = CountDownLatch(1)
+
+        Handler(Looper.getMainLooper()).postAtFrontOfQueue {
+            PixelCopy.request(
+                window,
+                bitmap,
+                { copyResult: Int ->
+                    Log.d(TAG, "PixelCopy result: $copyResult")
+                    if (copyResult != PixelCopy.SUCCESS) {
+                        Log.e(TAG, "Failed to capture screenshot")
+                        return@request
+                    }
+
+                    Log.e("BITMAP CAPTURED", bitmap.toString())
+                    val viewHierarchy = bitmapToVH[bitmap]
+
+                    if (viewHierarchy != null) {
+                        val canvas = Canvas(bitmap)
+                        viewHierarchy.traverse {
+                            if (it.shouldRedact && (it.width > 0 && it.height > 0)) {
```

Comment:
in theory all redaction could happen on the scaled down version
to  avoid manually scaling every `it.visibleRect`, simply set a matrix on the canvas based on the aspect ratio